[Mixer] added ValueTips for all faders and rotary knobs

oddtime opened the pull request:
Related to issue #1011
Basically the aim is to show the values if the user wants to know them

oddtime commented:
Also, I think that other rotary knobs in the mixer (humanize and swing parameters, FX volumes) could have` bUseValueTip` set `true ` to display the value on click...

oddtime commented:
For instruments and main fader:
It is sufficient to update the tooltip just in the method setVolume() 

For the master fader:
it is not sufficient. Updating the tooltip in the method faderChanged() is needed

searching why

oddtime commented:
In file `src/gui/src/Mixer/Mixer.cpp` in method `void Mixer::updateMixer()` line 606 ...
```
	// set master fader position
	float fNewVolume = pSong->get_volume();
	float fOldVolume = m_pMasterLine->getVolume();
	if (fOldVolume != fNewVolume) {
		m_pMasterLine->setVolume(fNewVolume);
	}
	m_pMasterLine->updateMixerLine();
```
so setVolume() for the master fader is called only if the masterVolume of the song (core) is different from the master fader (gui).

updateMixer() is called with frequency 20 times/seconds and everytime it calls setVolume() for every mixer fader, except the master fader 1. It seems to me a waste of operations, but maybe it is necessary.

Would you modify updateMixer() to set the master Fader (and so update its tooltip) every call or update the tooltip in method faderChange()?

oddtime commented:
Used the class `RotaryToolTip` to show the LCD displays while changing the FX sends

theGreatWhiteShark commented:
Hey @oddtime,

Could you merge this branch with current master? I will have a look

oddtime commented:
Ok done

theGreatWhiteShark commented on src/gui/src/Mixer/MixerLine.cpp:
```diff
@@ -270,6 +274,9 @@ float MixerLine::getVolume()
 	m_pFader->setValue( value );
+	char m_pVolumePos[7];
+	snprintf( m_pVolumePos, 6," %#.2f", value);
+	m_pFader->setToolTip( tr("Volume") + QString( m_pVolumePos ) );
```
Could you use the Qt string library instead of the more error prone C-style char array?

```c++
ref->setToolTip(tr("Volume") + QString( " %1" ).arg( value, 0, 'f', 2 ) );
```

theGreatWhiteShark replied:
Also, we should remove the tooltip of `m_pNameWidget` (in L. 162) since it probably annoy users using your new tooltips on a regular basis. In addition, it does not add much information anyway.

theGreatWhiteShark commented on src/gui/src/Mixer/MixerLine.cpp:
```diff
@@ -720,8 +737,11 @@ float MasterMixerLine::getVolume()
 
 void MasterMixerLine::setVolume( float value )
-{
+{		printf("arr\n");
```
is this a debugging artifact?

oddtime replied:
Ops ;)

theGreatWhiteShark reviewed (commented):
I like it. Especially as a user of a tiling window manager who usually only sees either the Mixer or the main window.

Two general remarks:
- How about making the tooltip of the faders sticky? Right now the tooltip disappears as soon as the fader is moved and the user has to either look at the status message in the main window or stop moving the fader, move the mouse of the fader, and the back on it again. If it would be shown regardless of the movement and constantly updated, this would be a great help.
- There is a huge redundancy the in Widgets classes. How about instead of adding some parts of the `Rotary` functionality into the `Knob` class, we make the later class inherit the former? AFAICS `Knob` is just an out-of-date copy&paste version of `Rotary` with a different width and single background. By inheriting from `Rotary` with a different set of variables during the initialization `Knob` would always have the latest features of `Rotary` and we can get rid off a lot of redundant code. (And it should be moved to the Rotary.cpp/h file)

oddtime commented:
> How about making the tooltip of the faders sticky?
Yes I agree

> There is a huge redundancy the in Widgets classes. How about instead of adding some parts of the Rotary functionality into the Knob class, 

You are right. I can try to do it, though I am not quite competent in this stuff.

theGreatWhiteShark commented:
>  I can try to do it, though I am not quite competent in this stuff.

Don't worry. If you have any questions just ask them straight away. In addition, since you already have a C background, I can recommend [A tour of C++](https://www.stroustrup.com/tour2.html) by Stroustrup to catch up with the modern version of C++. We already support C++17 and things will probably get more modern in times.

oddtime commented:
Thanks!

trebmuh commented:
@oddtime & @theGreatWhiteShark : do we want this to be in the 1.1 release?
(I'd vote yes)

oddtime commented:
Hi @trebmuh 
not necessarily, as @theGreatWhiteShark pointed the current code should be written in a better way to avoid redundancy with classes etc instead, and the fader tooltip is not optimal as it disappears on volume change. But I had no time to spend on programming in the last days and I wouldn't slow down the release

cme commented on src/gui/src/Mixer/MixerLine.cpp:
```diff
@@ -822,11 +826,11 @@ void MasterMixerLine::rotaryChanged( Rotary *pRef )
 	if ( pRef == m_pHumanizeTimeRotary ) {
 		pEngine->getSong()->setHumanizeTimeValue( fVal );
-		sMsg = tr( "Set humanize time parameter [%1]").arg( fVal, 0, 'f', 2 );
+		sMsg = tr( "Set hum. time parameter [%1]").arg( fVal, 0, 'f', 2 ); //not too long for display
```
I stared at this for an embarrassingly long period of time before realising I could just look to the left to see what 'hum.' is an abbreviation for.

Rather than shortening 'humanize', could we instead shorten 'parameter' to 'param'? This is a well established abbreviation that I think would make the message more readable?

cme commented:
Hey @oddtime 

> not necessarily, as @theGreatWhiteShark pointed the current code should be written in a better way to avoid redundancy with classes etc and the fader tooltip is not optimal as it disappears on volume change. But I had no time to spend on programming in the last days and I wouldn't slow down the release

We're otherwise ready to declare a String Freeze for translation, so if you don't feel comfortable reworking in time for 1.1 I'll tag this for the 1.2 milestone if that's okay with you.

oddtime commented:
@cme 
> We're otherwise ready to declare a String Freeze for translation, so if you don't feel comfortable reworking in time for 1.1 I'll tag this for the 1.2 milestone if that's okay with you.

Yes I agree, let's tag it for 1.2

trebmuh commented:
OK
When the 1.2 times will come, let me know if you need some testing here.

theGreatWhiteShark commented:
Hey @oddtime,

FYI: there is already an instance of a tooltip being visible the whole mouse interaction. In the Instrument Editor > Layers left click on the left or right border of a light blue rectangle indicating the velocity for which the layer should be triggered.

oddtime commented:
Thank you! :) I shall have a look definitely

ps: even if in this period I have less free time, I am currently reading A tour of C++ by Stroustrup that you have advised. I think I can be more helpful, and less tiresome for you reviewers, after learning the language better

theGreatWhiteShark commented:
> even if in this period I have less free time, I am currently reading A tour of C++ by Stroustrup that you have advised. I think I can be more helpful, and less tiresome for you reviewers, after learning the language better

Nice to hear. But don't worry about us devs too much. My first PRs here didn't require less time and attention and after all Open Source projects are also a place for learning about new things.

Take your time. I think it will be for at least a couple more weeks until we shift attention from the docs to the application again anyway.

oddtime commented:
@theGreatWhiteShark 
> How about instead of adding some parts of the Rotary functionality into the Knob class, we make the later class inherit the former?

What if we deprecate the knob class and introduce a TYPE_KNOB to the Rotary class? If the derived Knob class has not further members may this help to simplify the code even more?

theGreatWhiteShark commented:
> What if we deprecate the Knob class and introduce a TYPE_KNOB (or "TYPE_SMALL") to the Rotary class? If the derived Knob class has not further members may this help to simplify the code even more? Rotary class has currently TYPE_NORMAL and TYPE_CENTER

Sounds good. But please do not use too much times with the `Knob` class. There is a quite chance they will die when #1151 will be fixed (since they look odd  in an odd number and they imply to be read in a western-centric way from top left to bottom right while the plugins in the FX strip are just top to bottom).

oddtime commented:
Thanks!
I made some minor changes to the `Rotary` class to include that Knob objects.

Now the volume faders have sticky QToolTips (whose position follows the mouse global position), as the ones in Layers Panel.

I added the fine adjustment for the mouse wheel event on faders (pressing ctrl).

As the `Rotary` class missed a tooltip for mouse wheel event, I added a QToolTip for it, probably not the best looking way but I don't know how to hide the valuetip lcd after wheel event (since there is no release event).

theGreatWhiteShark commented:
Nice. When I'm finished with the docs (this should be a matter of days) I will have a look at it.

theGreatWhiteShark commented on src/gui/src/Widgets/Fader.cpp:
```diff
@@ -23,6 +23,7 @@
 
 #include "../Skin.h"
 #include "Fader.h"
+#include "Rotary.h"
```
How come you include the header of the Rotary class in here?

theGreatWhiteShark reviewed (commented):
Nice. I like your changes a lot!

oddtime commented on src/gui/src/Widgets/Fader.h:
```diff
@@ -29,6 +29,7 @@
 #include <QtWidgets>
 
 #include <core/Object.h>
+#include "Rotary.h"
```
Thanks. It was something related to the tooltip I added in the knob class before deprecation, so it should to be removed

oddtime commented on src/gui/src/Widgets/Rotary.cpp:
```diff
@@ -224,33 +234,44 @@ void Rotary::wheelEvent ( QWheelEvent *ev )
+		sprintf( tmp, "%#.2f", m_fValue );
+		
+		// Problem: ValueToolTip remains visible after wheel event
+		//m_pValueToolTip->showTip( mapToGlobal( QPoint( -38, 1 ) ), QString( tmp ) );
```
Sorry, I am not sure to follow, please could you explain it again? Some questions to understand better:
Do you mean writing on the object tooltip, the one that currently stores the description of the widget (e.g. "Filter cutoff"...)?
If so, should the description string ("Filter cutoff") remain visible when changing the value by mouse or wheel movement (eg "Filter cutoff 0.25")?
I think that when you change the value it is better to show just the number. But then the descriptive string should return in some way, and that is simple if one uses two dedicated tooltips of course.

> Also, being able to access the current value via tooltip would be both nice (as one does not have to move the widget in order to get its value)
Well yes, even if now you can click without move to show the value with the RotaryToolTip.

theGreatWhiteShark replied:
> If so, should the description string ("Filter cutoff") remain visible when changing the value by mouse or wheel movement (eg "Filter cutoff 0.25")?
I think that when you change the value it is better to show just the number. But then the descriptive string should return in some way, and that is simple if one uses two dedicated tooltips of course.

At first I would have said to have the description remain visible. But I really like the way you have done it for the fader and showing just the number makes a lot more clear what is happening.

oddtime replied:
Exactly, less information for the eye to filter

oddtime commented on src/gui/src/Widgets/Rotary.cpp:
```diff
@@ -224,33 +234,44 @@ void Rotary::wheelEvent ( QWheelEvent *ev )
+		//m_pValueToolTip->showTip( mapToGlobal( QPoint( -38, 1 ) ), QString( tmp ) );
+		
+		// useful but graphically incoeherent
+		QToolTip::showText( ev->globalPos(), QString( "%1" ).arg( m_fValue, 0, 'f', 2 ) , this );
```
Personally the graphical incoherence is not a problem for me, nor a priority at this point ( and I guess even for you ;) )
But yes I agree, I just prefer now to spend time in bugs and the new features like the float ticks and tuplets...
That pull req. was just for the sake of "reading the numbers", waiting for a new GUI eventually.

oddtime replied:
Of course if you already know how to use Qt it is fine to include the change here

theGreatWhiteShark replied:
> Personally the graphical incoherence is not a problem for me, nor a priority at this point ( and I guess even for you ;) )

Well, it is not on my todo list and could easily wait. But the GUI really needs an update (in the graphics -> port to SVG, the components -> using .ui files to represent e.g. the Instrument Editor instead of a solid bitmap with no internationalization, and the general UX -> like keyboard interaction, click&drag). And the fundamental widgets are a good place to start. A year ago I wouldn't have minded. But change of Colin to the Song and Pattern Editor showed me that tweaks in the UX can improve an old-fashioned app soo much it actually becomes pleasant to work with.

I think I will have another look at the tooltips to make them consistent for all widgets, try to make their look integrate better into the overall design of Hydrogen, and replace both the Rotary and Knob images by SVG.

So, if you want to spend time with other issues, I would suggest we merge this PR and I take over in a different one.

oddtime replied:
> So, if you want to spend time with other issues, I would suggest we merge this PR and I take over in a different one.

Perfect!

theGreatWhiteShark commented:
@oddtime could you resolve the merge conflict? Then we are ready to merge

oddtime commented:
Conflict resolved!
